### PR TITLE
feat: Sync common dependency versions between packages

### PR DESCRIPTION
**Pull request by lohnn**
**NOTE:** I'm adding this as a draft PR as docs and tests are not yet added and I haven't yet added feedback if the user turned on the flag but didn't provide the `common_packages.yaml`-file. But I wish for feedback on my solution anyways.

## Description

Added support for setting versions for common/shared dependencies that will be applied to all packages in a Melos monorepo project. Implements #94.

A new bootstrap config flag (`shareDependencies`) is added to start using this new feature, and then the dependency versions for `dependencies` and `dev_dependencies` in `common_packages.yaml` will be applied to all packages that contains said dependency.

## Type of Change

- [x] ✨ `feat` -- New feature (non-breaking change which adds functionality)
- [ ] 🛠️ `fix` -- Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ `!` -- Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 `refactor` -- Code refactor
- [ ] ✅ `ci` -- Build configuration change
- [x] 📝 `docs` -- Documentation
- [ ] 🗑️ `chore` -- Chore

### CODE REVIEW

**Comment by lohnn** [comment]
Another question: should Melos maybe sync the Dart and Flutter version in environment as well?

**Comment by vkammerer** [comment]
My two cents:
- Instead of a dedicated separate file, we could include these properties in `melos.yaml`. It is a feature of melos, so there is no reason to exclude it from the melos settings. At a later point, this could potentially also be a part of `pubspec.yaml` if https://github.com/invertase/melos/issues/472 is implemented.
- the name `shareDependencies` makes sense if the dependencies versions are taken from the root package's own dependencies, as suggested by @blaugold in https://github.com/invertase/melos/issues/94#issuecomment-1539961842, but since your implementation follows another approach, there is no "sharing" anymore. Instead, a more appropriate name could be `assignCommonDependencies` or similar.
- you could avoid calling https://github.com/invertase/melos/pull/526/files#diff-b5438310509b0fb6f04ef9776c2b9d3ec2a18fe0e0bc7165cdda954148ad9a99R240-R243 if `dependenciesUpdated == 0 && devDependenciesUpdated == 0`, (which will be the case most of the time, since `melos bootstrap` is likely to be run more often than common dependencies version are updated).
- There is a mixed usage of the terms "package" and "dependency", which is confusing. I may be wrong, but it looks to me that a "package" refers to a local package managed by melos, while a "dependency" refers to an entry of the "dependencies" and "dev_dependencies" fields. If that is indeed the case, there are several usages of "package" that should be replaced with "dependency":
  - if we wanted to go with a separate file, I guess it should be named "common_dependencies.yaml" rather than "common_packages.yaml"
  - `_updatePackages` -> `_updateDependencies`
  - `packagesToUpgrade` -> `dependenciesToUpgrade`
  - etc
- the naming of `dependenciesUpdated` and `devDependenciesUpdated` sound like booleans to me. I would suggest `updatedDependenciesCount` and `updatedDevDependenciesCount` to be more explicit.

But all these are details, compared to the more important design decision of whether we want to share the root package's dependencies or use separate fiels as you have done. As I have already expressed in https://github.com/invertase/melos/issues/94#issuecomment-1543268061 , I am more in favour of new fields, but it would be nice to hear @blaugold 's opinion on this design choice.

**Comment by lohnn** [comment]
I would also love to hear what @blaugold has to say as I should have some time over to continue with this PR next week.

**Comment by blaugold** [comment]
Sorry for not responding earlier! 😔 The past few weeks have been really busy.

@lohnn Thanks for working on this!

> Another question: should Melos maybe sync the Dart and Flutter version in environment as well?

Great idea. I suspect it's just a question of time until someone asks for this, especially with support for common dependencies.

@vkammerer Thanks for the detailed input. You convinced me that putting the common dependencies into the root pubspec is probably not a good idea. I feel like the next best option is to put them into `melos.yaml`. Like you, I don't really see a reason not to. We already maintain a JSON schema for `melos.yaml`, which can be used for autocompletion and validation. Extending that is much easier than introducing a new type of file.

Regarding naming, I would be fine with mirroring that of `pubspec.yaml` under the bootstrap command options:

```yaml
command:
  bootstrap:
    environment:
      ...
    dependencies:
      ...
    devDependencies:
      ...
```

**Comment by lohnn** [comment]
Heh, it's all good! I'm getting married soon and know exactly what you mean. Thankfully I can do some of this job during work hours (wish I could say the same for my own package, but we currently don't use it at work 😅)

> Great idea. I suspect it's just a question of time until someone asks for this, especially with support for common dependencies.

Then I'll add it to the TODO list.

Regarding putting the common dependencies in `melos.yaml` file I only have one gripe. The `melos.yaml` file is, at least in our project, already quite bloated clocking in at just short of 200 ilnes. Adding all the common dependencies would increase the size of this file even more.

On the flip side, we all use IDEs and can collapse parts of the file if we want to make it more readable, so if you still wish for the common dependencies (and version number) to live in the `melos.yaml` file, I'll update this PR accordingly :)

> Regarding naming, I would be fine with mirroring that of pubspec.yaml under the bootstrap command options:

I imagine you mean it should be:

```yaml
command:
  bootstrap:
    environment:
      ...
    dependencies: 
      ...
    dev_dependencies: # instead of devDependencies
      ...
```

**Comment by vkammerer** [comment]
@lohnn I would much prefer the simplicity of a few lines added to `melos.yaml` (which would be visible only when you view that file) than having to deal with yet another config file just for this purpose (which would always be visible when you browse your project in your code editor, which is to say almost all the time).

Moreover, the argument about it belonging to melos responsibilities and therefore belonging to `melos.yaml` still stands.

**Comment by lohnn** [comment]
Check! I'll take it and update the code accordingly 👍 (whenever I too get time 🤦‍♂️)

**Comment by lohnn** [comment]
And regarding tests for this work. I should look into that I guess?

**Comment by lohnn** [comment]
I have now added a test to validate that the bootstrap dependency sync runs as intended, and the PR is now ready to be reviewed.

**Comment by lohnn** [comment]
Ping: I would love for this PR to be reviewed.

**Comment by blaugold** [comment]
@lohnn Sorry for the delay! I'll review it tomorrow.

File: packages/melos/lib/src/commands/bootstrap.dart [hunk 1]
```diff
@@ -199,6 +214,127 @@ mixin _BootstrapMixin on _CleanMixin {
     }
   }
 
+  Future<void> _setSharedPreferencesInPackage(
```
Comment (blaugold) [inline packages/melos/lib/src/commands/bootstrap.dart:217]: Consider the name `_setSharedDependenciesForPackage`. `Environment` also specifies dependencies, just on SDKs not packages.

File: packages/melos/lib/src/commands/bootstrap.dart [hunk 2]
```diff
@@ -199,6 +214,127 @@ mixin _BootstrapMixin on _CleanMixin {
...
+    );
+
+    final updatedDevDependenciesCount = _updateDependencies(
+      pubspecEditor: pubspecEditor,
+      workspaceDependencies: devDependencies,
+      packageDependencies: package.pubSpec.devDependencies,
+      pubspecKey: 'dev_dependencies',
+    );
+
+    await writeTextFileAsync(
```
Comment (blaugold) [inline packages/melos/lib/src/commands/bootstrap.dart:247]: Can we avoid writing `pubspec.yaml` when we did not update anything?

File: packages/melos/lib/src/commands/bootstrap.dart [hunk 3]
```diff
@@ -199,6 +214,127 @@ mixin _BootstrapMixin on _CleanMixin {
...
+    final message = <String>[
+      if (updatedEnvironment) 'Updated environment',
+      if (updatedDependenciesCount > 0)
+        'Updated $updatedDependenciesCount dependencies',
+      if (updatedDevDependenciesCount > 0)
+        'Updated $updatedDevDependenciesCount dev_dependencies',
+    ];
+    if (message.isNotEmpty) {
+      logger.child(packageNameStyle(package.name), prefix: '').child(
+            message.join('\n'),
```
Comment (blaugold) [inline packages/melos/lib/src/commands/bootstrap.dart:261]: Removing the trailing comma in this line makes the whole log statement format a little nicer.

File: packages/melos/lib/src/commands/bootstrap.dart [hunk 4]
```diff
@@ -199,6 +214,127 @@ mixin _BootstrapMixin on _CleanMixin {
...
+
+  int _updateDependencies({
+    required YamlEditor pubspecEditor,
+    required Map<String, DependencyReference>? workspaceDependencies,
+    required Map<String, DependencyReference> packageDependencies,
+    required String pubspecKey,
+  }) {
+    if (workspaceDependencies == null) return 0;
+    // Filter out the packages that does not exist in package and only the
+    // dependencies that has a different version specified in the workspace.
```
Comment (blaugold) [inline packages/melos/lib/src/commands/bootstrap.dart:318]: does => do

has => have

File: packages/melos/lib/src/common/utils.dart [hunk 5]
```diff
@@ -321,6 +321,9 @@ String melosOverridesYamlPathForDirectory(String directory) =>
 String pubspecPathForDirectory(String directory) =>
     p.join(directory, 'pubspec.yaml');
 
+String pubspecCommonPathForDirectory(String directory) =>
+    p.join(directory, 'common_packages.yaml');
+
```
Comment (blaugold) [inline packages/melos/lib/src/common/utils.dart:326]: This is not used anywhere.

File: docs/commands/bootstrap.mdx [hunk 6]
```diff
@@ -84,6 +84,45 @@ example:
...
+dependencies:
+  collection: ^1.17.0
+  uni_links2:
+  uni_links_macos:
+    git: https://github.com/SamJakob/uni_links_macos.git
+
+dev_dependencies:
+  build_runner: ^2.3.3
+# ...
+```
```
Comment (blaugold) [inline docs/commands/bootstrap.mdx:124]: This looks outdated since it talks about `common_packages.yaml`.

**Review by blaugold — APPROVED** [review]
LGTM

Thanks! ❤️

**Comment by feinstein** [comment]
I am trying to find this mechanism in the docs website, but there I can only see a `dependencies` section in `bootstrap`. Has the `common_dependencies.yaml` file being removed?

**Pull request merged**